lib/src/diff.rs

```diff
@@ -471,14 +471,14 @@ impl<'input> Diff<'input> {
         diff
     }
 
-    pub fn unrefined<T: AsRef<[u8]> + ?Sized + 'input>(
+    pub fn unrefined<T: AsRef<BStr> + ?Sized + 'input>(
```

Comment:
Naively, does `BStr` satisfy `AsRef<[u8]>`? Can it be built from `AsRef<u8]>`? If so, maybe it's better to keep the type as `AsRef<[u8]>`?

I do find `&*` confusing. My understanding of Rust is insufficient to understand exactly what happens there. I might, of course, look it up.

Reply:
I read up on the `bstr` crate, and it *seems* like it's supposed to work well with `[u8]` arguments everywhere. For example, in https://docs.rs/bstr/latest/bstr/index.html#bstr-in-public-apis it says that you *could* use BStr in public APIs, but there's usually no need for that.

I could easily be missing something important, though.

Reply:
`[u8]` also works. I choose `BStr` just because it's shorter and seems to explain the API better.

Reply:
I still haven't fully figured out what's going on, but I think we should avoid `&*` if we can. If `[u8]` vs `BStr` makes a difference to that, I'd prefer the former. If not, either is fine.

Reply:
Also, we're actually talking about `AsRef<[u8]>` vs `AsRef<BStr>`, and more types from other crates are likely to implement the former. So, `AsRef<[u8]>` seems better.

Reply:
Maybe I should send an upstream PR that adds `AsRef<BStr> for Vec<u8>`.

Reply:
Just to be clear, I'm 100% behind using BStr internally in whichever types this makes debug-printing better for.

If it makes a difference, I'd be behind using it for other purposes as well. AFAICT based on what I read so far, the debug string is the only actual meaningful difference between `BStr` and `[u8]` because of the traits the `bstr` crate adds to `[u8]`, but I might be missing something.

>  I choose BStr just because it's shorter and seems to explain the API better.

I think Rust already conditions programmers to think "bytestring" when they see `&[u8]`, since that's the type of `b"bytestrings"`. So, I don't think there is much advantage to clarity in using BStr over it.

> Maybe I should send an upstream PR that adds AsRef<BStr> for Vec<u8>.

That wouldn't hurt, of course, but I'd still think that using `AsRef<[u8]>` would be slightly better. At that point, the difference would of course matter less.

Reply:
Ok, I don't have strong opinion, so switched to `[u8]`.